vortex-file/src/scan/mod.rs

```diff
@@ -0,0 +1,286 @@
+                );
+                Some(RowMask::new(filter_mask, row_range.start))
+            })
+            .collect_vec();
+
+        let result_dtype = projection.return_dtype(layout_reader.dtype())?;
+
+        // Create a future to process each row split of the scan.
+        let array_futures: Vec<_> = row_masks
+            .into_iter()
+            .enumerate()
+            .map(|(_i, row_mask)| {
+                let row_range = row_mask.begin()..row_mask.end();
+
+                let approx_filter_eval = filter
+                    .as_ref()
+                    .map(|expr| layout_reader.pruning_evaluation(&row_range, expr))
+                    .transpose()?;
+                let exact_filter_eval = filter
+                    .as_ref()
+                    .map(|expr| layout_reader.filter_evaluation(&row_range, expr))
+                    .transpose()?;
+                let project_eval = layout_reader.projection_evaluation(&row_range, &projection)?;
+
+                Ok::<_, VortexError>(instrument!("split", [split = _i], async move {
+                    let mut mask = row_mask.filter_mask().clone();
+                    if mask.all_false() {
+                        return Ok(None);
+                    }
+
+                    if let Some(approx_filter_eval) = approx_filter_eval {
+                        // First, we run an approximate evaluation to prune the row range.
+                        log::debug!("Pruning row range {:?}", row_range);
+                        mask = approx_filter_eval.invoke(mask).await?;
+                        if mask.all_false() {
+                            return Ok(None);
+                        }
+                    }
+
+                    if let Some(exact_filter_eval) = exact_filter_eval {
+                        // Then, we run the full evaluation.
+                        log::debug!("Filtering row range {:?}", row_range);
+                        mask = exact_filter_eval.invoke(mask).await?;
+                        if mask.all_false() {
+                            return Ok(None);
+                        }
+                    }
+
+                    log::debug!("Projecting row range {:?}", row_range);
+                    let mut array = project_eval.invoke(mask).await?;
+                    if self.canonicalize {
+                        log::debug!("Canonicalizing row range {:?}", row_range);
+                        let mut builder = builder_with_capacity(array.dtype(), array.len());
+                        array.append_to_builder(builder.as_mut())?;
+                        array = builder.finish();
+                    }
+
+                    Ok(Some(array))
+                }))
+            })
+            .try_collect()?;
+
+        // Spawn the array futures onto the executor and buffer some number of row splits.
+        let task_executor = self
+            .task_executor
+            .unwrap_or_else(|| TaskExecutor::Threads(ThreadsExecutor::default()));
```

Comment:
nit - if this isn't the `Default` impl, it should be

Reply:
Yes, although I agree we should defer the construction since this spins up threads. Discussed in person that the new default won't be threads model, so will FLUP.